### PR TITLE
Is this Turing complete? May need set command

### PR DESCRIPTION
**Pull request by bparke37**

### CODE REVIEW

**Comment by bparke37** [comment]
In order for a language to be Turing complete it must be able to simulate a Turing machine. The biggest issue I see in doing that with this language, is that there is no way to enter data into a line without it coming from an external source. This commit fixes that by adding a set data command.

**Comment by AZHenley** [comment]
Why isn't an external source sufficient? Is memory not external too?

**Comment by bparke37** [comment]
From a Turing Machine perspective, the tape alphabet and the input alphabet are not constrained to be the same. With hofstadter in it's current state they are constrained to be the same. This means that there is no way to have the program output "Hello World!" without it first reading in "Hello World!" which would be possible by a Turing Machine. 

From a processor perspective, the processor can write any amount of arbitrary data to anywhere in memory without having to first read that data in. With hofstadter in it's current state, those sort of actions are impossible. 

In essence this means that hofstadter lacks the capability to create data, only manipulate data. In this way it's closer to SQL in it's use cases than it is to Java. And even then, it would lack certain features of SQL. 

This commit hopes to remedy all of that by adding in a set data command, which will bring it closer to a Turing Complete Language.

**Comment by AZHenley** [comment]
Lambda calculus is Turing complete and it doesn't have any form of alphabet and no variables.

**Comment by bparke37** [comment]
Lambda Calculus is only Turing complete because of mechanisms like [Church Encodings](https://en.wikipedia.org/wiki/Church_encoding) which allow lambda calculus expressions to become themselves pieces of data to be manipulated arbitrarily.

Fundamentally here is my line of thought:

**Premise Zero:**
Input is any data which is not created by the language/machine itself. I.E. Data gathered from an external source.
**Premise One:**
Turing completeness requires that your language can do at least anything that a Turing machine can do.
**Premise Two:**
Turing machines can output data that has not been input to them.
**Premise Three:**
Hofstadter cannot output data that has not been input into it. 

**Conclusion:**
Hofstadter is not Turing complete.

**Solution:**
Create a function to allow Hofstadter to output data that has not been input to the language. 

**Proof of Premise Three:**
Hofstadter has 8 commands. All internal Hofstadter data starts as null.
1. HTTP requests get data through input. 
2. Regex performs manipulations on already stored data.
3. File I/O can get data through input.
4. Console I/O can get data through input.
5. Conditional statements can only check data.
6. ^
7. The swap statement can only swap data.
8. The concatenate statement can only concatenate existing data. 

**Therefore:**
No command within Hofstadter allows for the creation of data by the language itself. 

**Conclusion:**
Hofstadter cannot output data that has not been input into it.

Either my above rationalization is wrong, or Hofstadter is not Turing complete.

**Comment by AZHenley** [comment]
This is all very interesting! But the distinction you are making between data, files, external sources, and alphabet seem arbitrary. It is not as if Turing machines know what a code file or data file is. Why can't files be the language's alphabet in the same way that .data is for assembly programs or literals in C are? In that case, the input and output alphabets are all possible values.

Would it help if Hofstadter explicitly defines the alphabet as being all files?

**Comment by bparke37** [comment]
Let me try to explain it this way:

A Turing Machine is described by a 5 tuple consisting of:
1. A set of states.
2. An input alphabet
3. A tape alphabet
4. A state transition function
5. And a start state.

The Turing machine is considered to be a "head" running on an infinite tape of cells which can each contain one symbol.

We can define input as the data which is on the tape before the machine starts running. Since the tape is equivalent to memory, the input to the program is all memory that your program has access to when your program's main function begins to execute (The main function being equivalent to the start state of the Turing machine). This includes both on your local machine and the internet as a whole. 

Hofstader cannot write to (memory/ the tape) without having to read from (memory/ the tape).
A Turing machine **can** write to (memory/ the tape) without having to read from (memory/the tape). 

In this way all the data of a program that exists in global memory before the program begins to execute (Such as .data in assembly or some literals in C) are considered part of the input, but data that is written to the stack at runtime is considered part of the output. 

In other words, a Turing machine can write data to the stack. Holfstader can only copy. 

Now I think I was wrong in saying that the tape alphabet would be limited to the input alphabet. If we think of the tape as storing not just binary data but objects which we represent using binary data, then I think it would actually be all possible concatenations and regex operations on the input alphabet. I'm unsure if this is equivalent or not to a Turing machine's data manipulation abilities.

**Comment by AZHenley** [comment]
> Hofstader cannot write to (memory/ the tape) without having to read from (memory/ the tape). A Turing machine can write to (memory/ the tape) without having to read from (memory/the tape).

This is a really interesting point.

I find it funny that even though this language might not be Turing complete, that it is far more useful than something like Brainf**k, which is! 

I intentionally didn't add a set operation since it is way more fun that way (it is an esolang after all 😄).  I think you've done a great job making a case that this isn't Turing complete though. Please add anymore rationale you come up with for/against the equivalence to a Turing machine.

**Comment by bparke37** [comment]
I had another thought while going through this. I think using strings as data values will mess with the Turing completeness. Specifically since strings are null terminated.

If hofstader was a Turing complete language it should be able to write any sequence of binary data that a Turing machine can. Yet, every sequence of data in hofstader will be written with a null terminating byte, while a Turing machine can write data that doesn't end in a null terminating byte. 

Just had another thought, a Turing machine should be able to write any length of bits to the tape, while hofstader will only be able to write 8 bit increments.

**Comment by AlexanderKrneta** [comment]
I was just about to somehow notify @AZHenley that this language is Turing complete. But wow, I just found this thread. I am writing a sort of compiler right now, to generate a Hofstadter file from a TM file, that should simulate a TM with the Hofstadter code. I should be done with it soon.

**Comment by jwg4** [comment]
I think that these are trivialities. Turing completeness does not require that a machine can produce the same outputs as a Turing machine. It is perfectly valid for there to be a translation step which decodes the output produced by the given machine into the corresponding output produced by a Turing machine.

A machine which can emulate a Turing machine with the tape alphabet limited to 0 and 1 can also emulate an arbitrary Turing machine without this restriction. This is because Turing machines of these two types can emulate each other. Clearly, for them to do so necessitates a transformation between tapes with only 0 and 1, and tapes with some other alphabet.

This is made clear by the fact that for example, Minesweeper has been shown to be Turing complete. Minesweeper has only around a dozen cell states (hidden bomb, flagged hidden bomb, hidden empty, numbers 0-8, etc). How does Minesweeper emulate Turing machines with more than this number of tape states? Simply in the same way it emulates 0-1 Turing machines - using a possibly complicated rule for translating between (collections of) Minesweeper cell states and Turing machine states (of whatever tape alphabet). In no case is there a direct 1-1 correpondence between Minesweeper cells and Turing machine tape cells.

Suppose that Hofstadter can act as a Turing machine with tape output 0 and 1, and that this emulation involves some translation between Hofstadter output such as "verbatim contents of the website www.google.com in file N <-> 0 at tape position N" and "verbatim contents of the website goatse.cz in file N <-> 1 at tape position N". This would be enough to make Hofstadter Turing complete. This also means that a Hofstadter program + two input files with different contents can emulate a Turing machine. Since a Hofstadter machine can write those two files before it does anything else, there is a basic Hofstafter program which can emulate any Turing machine.

**Comment by bparke37** [comment]
@jwg4 I agree that my last comment was wrong. They were sort of spur of the moment thoughts that I hadn't actually given much thought to. I never corrected them because I didn't think anyone would actually care one and a half years into the future.

You are correct in saying that the Turing machine doesn't need to actually output an arbitrary sequence of bits to be Turing complete. As long as a complete translation exists between the tape alphabets they can be considered equivalent. Thus, as long as a program can write two different pieces of data to memory, those can be mapped to zero and one and it's tape alphabet can be translated to binary. 

I believe you haven't actually proven my comment before last as false though. You state "This also means that a Hofstadter program + two input files with different contents can emulate a Turing machine". I don't disagree with this, but that wasn't the point I was making. I was making the point that a Hofstadter program in isolation cannot produce data in any form (at the time of making the commit; I'm not sure how the language has evolved since then). I stated above that: 

"We can define input as the data which is on the tape before the machine starts running ... This includes both on your local machine and the internet as a whole ... Hofstadter cannot write to (memory/ the tape) without having to read from (memory/ the tape). A Turing machine can write to (memory/ the tape) without having to read from (memory/the tape)".

I believe that this still holds. To put it another way, say that there was no input to a Hofstadter program, i.e. there was no data stored across the entire internet or on your local machine. If Hofstadter, the programming language only, was Turing complete, it should be able to output any data a Turing machine could output, given a complete translation exists between the Turing machine's tape alphabet and the tape alphabet Hofstadter has access to. Since Hofstadter cannot write data to memory that was not already within memory, it does not satisfy this requirement and is not Turing complete. 

I agree that Hofstadter would be Turing complete if it could always have some amount of data available to it at the start of the program for it to use to write to memory. Since that was not the case at the time of this commit though (and still might not be), Hofstadter is not Turing complete.

**Comment by jwg4** [comment]
I think you are still well off the mark in understanding what Turing
completeness is.

Your quote:
> "We can define input as the data which is on the tape before the machine
starts running ... This includes both on your local machine and the
internet as a whole ... Hofstadter cannot write to (memory/ the tape)
without having to read from (memory/ the tape). A Turing machine can write
to (memory/ the tape) without having to read from (memory/the tape)".

This is wrong on two counts.

First of all I believe you have misunderstood how Hofstadter source code
works. Writing constants in the source code, which can be written to files,
without having to read them first, is a perfectly normal part of the
language.

Secondly, and much more fundamentally, your statement "we can define
input..." shows a much more fundamental misconception about tiring
completeness. Turing completeness of a given machine/language/system means
that something exists, a mapping from Turing machines to instances of that
system. Lack of Turing completeness means that no such mapping can exist.

If you want to show that no valid mapping of Turing machines to Hofstadter
programs exists, you are not free to define what the input of the
Hofstadter program is. You have to allow that the person producing the
mapping can choose exactly which parts of the Hofstadter system (code +
language) correspond to the input tape, the input alphabet, and the
internal states of the 'tape reader'.

You would have to show that any conceivable mapping would fail to replicate
some aspect of one or more Turing machines. This argument cannot hope to
succeed when you choose yourself which parts of the Hofstadter system
correspond to input tape. This is what my example about minesweeper
illustrated – was it clear to you?

If, in my hypothetical construction of a Hofstadter system corresponding to
every Turing machine, the Hofstadter systems always start by writing 0 and
1 to two files called '0.txt' and '1.txt', that is a perfectly good
reduction of Turing complete problems to Hofstadter systems. It would be
equally good if I said "I also need the files '0.txt' and '1.txt' to be
available on the machine before I start. In fact, any finite set of
constants or seed data is perfectly valid, no different from if you
specified that a CPU has to start turned on with 0 (or any other value) in
its instruction register. (The restriction to finiteness means that you
can't cheat by asking for an infinite lookup table where the (N,M,K)th
entry shows the state of the Nth Turing machine with the Mth possible input
at the Kth step.)

If you look at some examples of how Turing completeness is shown, I think
you will quickly see that this principle is correct. On the other hand, if
you aren't convinced, there is code in this repo giving a reduction to
Hofstadter of every Turing machine. Can you show a Turing machine for which
it fails? Trying to do so might clarify your ideas about this language not
being Turing complete.

**Comment by AlexanderKrneta** [comment]
Hello! I did finish writing the compiler from a TM file to a Hofstadter file. with outputs/animations and all.
Because a functioning Turing Machine can be emulated with this language, and Turing machines are Turing complete, the Hofstadter Language is Turing Complete.

https://github.com/AZHenley/hofstadter/tree/main/examples/tmcompiler

Also, Turing completeness means that it could solve any computational problem. Though you don't need to emulate a Turing machine to prove Turing completeness, you can ALWAYS emulate a Turing machine on any TC language.

**Comment by bparke37** [comment]
@jwg4 
@AlexanderKrneta

I appreciate both of your responses and the time you both put into them. Unfortunately, I don't really have the time to do the same. I still don't agree that the Hofstadter language is Turing complete, but I'm not going to challenge your points any further.

**Comment by AlexanderKrneta** [comment]
> @jwg4 @AlexanderKrneta
> 
> I appreciate both of your responses and the time you both put into them. Unfortunately, I don't really have the time to do the same. I still don't agree that the Hofstadter language is Turing complete, but I'm not going to challenge your points any further.

It emulates a Turing machine. There is nothing to challenge.
